### DISCUSSION TITLE
Modifications from table_name to schema_name

### DISCUSSION THREAD

**Discussion by TonyGravagno** [araera111/mysql-to-zod#27] [Ideas]
I'd like to add options to the config file that will allow us to apply text and RegExp substitutions on a table name to generate its schema name. I already do this extensively and it helps a lot.

(The following is a rough guideline, not a final spec.)

### For ReadMe:
> This option consists of an array of substitution pairs. Each pair represents a From/To replacement for the table name, with the result applied to the new type name. If the From value begins and ends with "/", it will be processed as a regular expression. The To values do not need slashes.

### Examples:

| Table Name | pre-Pascal Schema Name | post-Pascal Schema Name | Replacement Array |
| ------------- | ------------- | ------------- | ------------- |
| foo_bar_baz | foo_blah_baz  | FooBlahBaz  | `[ '_bar_' , '_blah_' ]` |
| xcustomers  | customers_xref  |  CustomersXref  | `[ '/^x(.*)/' , '$1_xref' ]` |
| wp_users  | users  | Users  | `[ '/^wp_/' , '' ]` |
| auditlog, translog  | audit_log, trans_log  | AuditLog, TransLog  | `[ 'log' , '_log' ]` |
| foo_bar_baz  | baz_bar_foo  | BazBarFoo | `[ '/^(.*)_(.*)_(.*)$/' , '$3_$2_$1' ]` |

mysqlToZod.config.js : 

```js
replacements: [
   [ '_bar_' , '_blah_' ],
   [ '/^x(.*)/' , '$1_xref' ],
   [ '/^wp_/' , '' ],
   [ '/^(.*)_(.*)_(.*)$/' , '$3_$2_$1' ],
],
```

**Comment by araera111** [comment]
I wrote tests and functions during my lunch break at work.
It seems easy to implement this option, so I would like to include it in the next version.

```ts
type ReplaceTableNameParams = {
  tableName: string;
  replacement: string[];
};

export const replaceTableName = ({
  tableName,
  replacement,
}: ReplaceTableNameParams): string => {
  const [before, after] = replacement;
  /* if replacement[0]or[1] undefined -> return original tableName */
  if (isNil(before) || isNil(after)) return tableName;

  /* if notRegexp -> replace */
  /* use match? invalid regexp -> return original tablename */
  if (!before.startsWith("/") && !after.endsWith("/"))
    return tableName.replace(before, after);

  /* if regexp -> replace */
  const regex = new RegExp(before.slice(1, -1));
  return tableName.replace(regex, after);
};
```

```ts
describe("replaceTableName", () => {
  it("case1 match", () => {
    const tableName = "foo_bar_baz";
    const replacement: string[] = ["_bar_", "_blah_"];
    const result = "foo_blah_baz";
    expect(replaceTableName({ tableName, replacement })).toBe(result);
  });

  it("case2 not match", () => {
    const tableName = "foo_bar";
    const replacement: string[] = ["_bar_", "_blah_"];
    const result = "foo_bar";
    expect(replaceTableName({ tableName, replacement })).toBe(result);
  });

  it("case3 regexp", () => {
    const tableName = "xcustomers";
    const replacement: string[] = ["/^x(.*)/", "$1_xref"];
    const result = "customers_xref";
    expect(replaceTableName({ tableName, replacement })).toBe(result);
  });

  it("case4 regexp2 ^wp_", () => {
    const tableName = "wp_users";
    const replacement = ["/^wp_/", ""];
    const result = "users";
    expect(replaceTableName({ tableName, replacement })).toBe(result);
  });

  it("case5 regexp3 log", () => {
    const tableName = "auditlog";
    const replacement = ["log", "_log"];
    const result = "audit_log";
    expect(replaceTableName({ tableName, replacement })).toBe(result);
  });

  it("case6 regexp4 log", () => {
    const tableName = "foo_bar_baz";
    const replacement = ["/^(.*)_(.*)_(.*)$/", "$3_$2_$1"];
    const result = "baz_bar_foo";
    expect(replaceTableName({ tableName, replacement })).toBe(result);
  });
});

```

If multiple conditions are met, should the second process be done after the first?　Or should it be completed when the first process is finished?

```js
replacements: [
   [ '_bar_' , '_blah_' ],
   [ '/^(.*)_(.*)_(.*)$/' , '$3_$2_$1' ],
]
```
```js
tableName = "foo_bar_baz" -> foo_blah_baz -> baz_blah_foo ?
or foo_bar_baz -> foo_blah_baz ?
```

Is it correct that replace replaces the first word, or if I want to replaceAll, do I put a g in regexp?
Or should it be replaceAll for the word replace?

```js
['_bar_', '_blah_']
foo_bar_baz_bar_baz  -> ??
```

**Comment by TonyGravagno** [comment]
I think it needs to be up to the user of this tool to craft the replacement parameters to achieve the desired results. For example, I have a rule in my code that converts specific keywords to _keyword. Then when the camelCase / PascalCase is invoked it converts the newly inserted underscores into capitals:

**Tables:** commsmailingaddress, ipaddress, auditlog

**Replacements:**
mailing, _mailing
address, _address
ipaddress, _i_p_address
log, _log

**Schema:** CommsMailingAddress, IPAddress, AuditLog

Note, I have a config option to PascalCase rather than camelCase so that schema names are capitalized.

Related to order of operations, consider the replacements related to address. As-written that will not produce the desired _i_p_address, because ipaddress will be converted to ip_address before the check for ipaddress. The result will remain ip_address, which becomes IpAddress (which is pedantically correct but personally not my preference).

To address this, it's necessary to reverse and modify the rules:
ipaddress, _i_paddress
address, _address

Now ipaddress becomes _i_paddress, and that becomes _i_p_address, resulting in IPAddress.

A developer might need to do more with this, like converting a string to something unique like $addr$ and then modifying that later in the replacement loop ... which is very typical in operations like this.

**Comment by araera111** [comment]
I will close this issue as well, as I believe it has been resolved with your help.
